src/System.Management.Automation/engine/Subsystem/SubsystemManager.cs

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Management.Automation.Internal;
+
+#nullable enable
+
+namespace System.Management.Automation.Subsystem
+{
+    /// <summary>
+    /// Class used to manage subsystems.
+    /// </summary>
+    public static class SubsystemManager
+    {
+        private static readonly ReadOnlyCollection<SubsystemInfo> s_subsystems;
+        private static readonly ReadOnlyDictionary<Type, SubsystemInfo> s_subSystemTypeMap;
+        private static readonly ReadOnlyDictionary<SubsystemKind, SubsystemInfo> s_subSystemKindMap;
```

Comment:
Why do we use ReadOnly* types? It seems to disallow dynamic subsystem registration.

Reply:
Those fields are collections that won't change after initialization. They contain the metadata of each subsystem and the mapping between type/subysstemkind with the subsystem metadata. Registration happens to each `SubsystemInfo`.